istioctl/cmd/webhook.go

```diff
@@ -0,0 +1,714 @@
+# Display the webhook configuration of Galley and Sidecar Injector
+istioctl experimental post-install webhook status --validation --validation-config istio-galley 
+  --injection --injection-config istio-sidecar-injector
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := opts.Validate(); err != nil {
+				return err
+			}
+			client, err := createInterface(kubeconfig)
+			if err != nil {
+				return fmt.Errorf("err when creating Kubernetes client interface: %v", err)
+			}
+			validationErr, injectionErr := displayWebhookConfig(client, opts)
+			if validationErr != nil || injectionErr != nil {
+				return fmt.Errorf("error when displaying webhook configurations. validation err: %v. injection err: %v",
+					validationErr, injectionErr)
+			}
+			return nil
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.BoolVar(&opts.validationWebhook, "validation", true, "Display "+
+		"the validating webhook configuration.")
+	flags.BoolVar(&opts.injectionWebhook, "injection", true, "Display "+
+		"the injection webhook configuration.")
+	flags.StringVar(&opts.validatingWebhookConfigName, "validation-config", "istio-galley",
+		"The name of the ValidatingWebhookConfiguration to display.")
+	flags.StringVar(&opts.mutatingWebhookConfigName, "injection-config", "istio-sidecar-injector",
+		"The name of the MutatingWebhookConfiguration to display.")
+
+	return cmd
+}
+
+// Create the validatingwebhookconfiguration
+func createValidatingWebhookConfig(k8sClient kubernetes.Interface,
+	config *v1beta1.ValidatingWebhookConfiguration) (*v1beta1.ValidatingWebhookConfiguration, error) {
+	var whConfig *v1beta1.ValidatingWebhookConfiguration
+	var err error
+	client := k8sClient.AdmissionregistrationV1beta1().ValidatingWebhookConfigurations()
+	_, err = client.Get(config.Name, metav1.GetOptions{})
+	if err == nil {
+		fmt.Printf("update webhook configuration %v\n", config.Name)
+		whConfig, err = client.Update(config)
+	} else {
+		fmt.Printf("create webhook configuration %v\n", config.Name)
+		whConfig, err = client.Create(config)
+	}
+	return whConfig, err
+}
+
+// Create the mutatingwebhookconfiguration
+func createMutatingWebhookConfig(k8sClient kubernetes.Interface,
+	config *v1beta1.MutatingWebhookConfiguration) (*v1beta1.MutatingWebhookConfiguration, error) {
+	var whConfig *v1beta1.MutatingWebhookConfiguration
+	var err error
+	client := k8sClient.AdmissionregistrationV1beta1().MutatingWebhookConfigurations()
+	_, err = client.Get(config.Name, metav1.GetOptions{})
+	if err == nil {
+		fmt.Printf("update webhook configuration %v\n", config.Name)
+		whConfig, err = client.Update(config)
+	} else {
+		fmt.Printf("create webhook configuration %v\n", config.Name)
+		whConfig, err = client.Create(config)
+	}
+	return whConfig, err
+}
+
+// Enable webhook configurations
+func enableWebhookConfig(client kubernetes.Interface, opt *enableCliOptions) (error, error) {
+	var validationErr, injectionErr error
```

Comment:
Use https://github.com/hashicorp/go-multierror to return a single error with zero or more combined errors. The caller prints an error and exits if either fail. Similar comment applies in a further other locations in this file.

Reply:
If using multierror, it is not straightforward to know which part (validation or injection) has error. In some cases, a user/a test/a caller may want to know which part has error.

Reply:
`enableValidationWebhookConfig` and `enableMutationWebhookConfig` both wrap their errors messages with validation/injection verbiage.

Reply:
From the error message aspect, multierror is fine in expressing an error message consisting of multiple errors. But compared with returning two errors directly, multierror is not as straightforward in determining which part has error. The following is an example that illustrates returning two errors in foo() is more straightforward to know which part of foo() failed in a test.

err1, err2 := foo()
if expectPart1Succeed && err1 !=nil {
    print("Part 1 should succeed but failed in the test")
}
if expectPart2Succeed && err2 !=nil {
    print("Part 2 should succeed but failed in the test")
}

Reply:
This is true. However, this function is not exported (internal) with one caller and it doesn't handle  errors this way. It only prints either/both errors.

Reply:
Done.